File: WEEK_4_C/Q8.c
#include <stdio.h>
#include <stdlib.h>

struct Node{ int data; struct Node* next; };

int main(){
    struct Node *head=NULL,*last=NULL,*t;
    for(int i=1;i<=4;i++){
        t=(struct Node*)malloc(sizeof(struct Node));
        t->data=i;
        if(!head){ head=t; head->next=head; last=head; }
        else{ last->next=t; t->next=head; last=t; }
    }
    int key=2;
    struct Node *p=head,*prev=last;
    do{
        if(p->data==key){
            if(p==head) head=head->next;
            prev->next=p->next; free(p); break;
        }
        prev=p; p=p->next;
    }while(p!=head);
    t=head;
    do{ printf("%d ",t->data); t=t->next; }while(t!=head);
    return 0;
}

File: WEEK_4_Python/Q8.py
class Node:
    def __init__(self, data):
        self.data, self.next = data, None

def print_list(head):
    cur = head
    while cur:
        print(cur.data, end=" ")
        cur = cur.next
    print()

def pairwise_swap(head):
    cur = head
    while cur and cur.next:
        cur.data, cur.next.data = cur.next.data, cur.data
        cur = cur.next.next
    return head

# Example
head = Node(1); head.next = Node(2); head.next.next = Node(3); head.next.next.next = Node(4)
head = pairwise_swap(head)
print_list(head)   
